File: frontend/js/views/yield_to_call.js
// Yield to Call (YTC) — for callable bonds, the issuer can redeem at a stated
// price on stated dates. YTC is the IRR if held to first call; YTM if held to
// maturity; yield-to-worst is the lower of the two. Computation runs server-side
// via /calc/yield-to-call (traderview-core::yield_to_call) — a faithful port of
// the former client-side bisection solver, Python-pinned and unit-tested.
// Class-based styling (no inline styles) for release-WebKit correctness.

import { api } from '../api.js';
import { applyUiI18n } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import { esc } from '../util.js';
import * as enh from '../calc_enhance.js';

const fmt = (n, d) => (n == null || !Number.isFinite(Number(n)) ? '—' : Number(n).toLocaleString(undefined, { minimumFractionDigits: d, maximumFractionDigits: d }));
const VIEW = 'yield-to-call';
let lastReport = null;
let lastBody = null;

export async function renderYieldToCall(mount, _state) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.yield_to_call.title">// YIELD TO CALL · YTC vs YTM vs YTW</span></h1>
        <p class="muted small" data-i18n-html="view.yield_to_call.intro">
            Callable bonds: issuer can redeem at a stated price on stated dates.
            <strong>YTC</strong> is the IRR assuming the bond is called at the
            first call date. <strong>YTM</strong> assumes held to maturity.
            <strong>Yield-to-worst (YTW)</strong> is the lower of YTM and any
            YTC — that's the yield you should evaluate against, since the
            issuer will only call when it benefits them.
        </p>
        <div class="chart-panel">
            <div class="re-grid">
                <label><span class="muted small">Current price (% of par)</span>
                    <input type="number" id="ytc-price" step="0.1" min="20" max="200" value="106.5"></label>
                <label><span class="muted small">Coupon rate %/yr</span>
                    <input type="number" id="ytc-coupon" step="0.125" min="0" max="20" value="5.5"></label>
                <label><span class="muted small">Years to maturity</span>
                    <input type="number" id="ytc-mat" step="0.25" min="0.5" max="40" value="10"></label>
                <label><span class="muted small">Years to first call</span>
                    <input type="number" id="ytc-call" step="0.25" min="0.5" max="40" value="3"></label>
                <label><span class="muted small">Call price (% of par)</span>
                    <input type="number" id="ytc-callprice" step="0.5" min="80" max="120" value="100"></label>
                <label><span class="muted small">Payments / yr</span>
                    <select id="ytc-freq">
                        <option value="1">Annual (1)</option>
                        <option value="2" selected>Semi-annual (2)</option>
                        <option value="4">Quarterly (4)</option>
                    </select></label>
            </div>
            <button class="btn btn-sm primary" id="ytc-run">⚡ Compute</button>
            <div id="ytc-tools" class="ce-toolbar"></div>
            <div id="ytc-result" class="re-result"></div>
        </div>
    `;
    applyUiI18n(mount);

    const num = (id) => parseFloat(mount.querySelector(id).value) || 0;
    const readBody = () => ({
        price_pct: num('#ytc-price'),
        coupon_rate_pct: num('#ytc-coupon'),
        years_to_maturity: num('#ytc-mat'),
        years_to_call: num('#ytc-call'),
        call_price_pct: num('#ytc-callprice'),
        coupons_per_year: parseInt(mount.querySelector('#ytc-freq').value, 10) || 2,
    });
    const compute = async () => {
        const body = readBody();
        try {
            const r = await api.calcYieldToCall(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body);
        } catch (err) {
            showToast(err.message || 'Could not compute the yields.', { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#ytc-tools'), {
        viewId: VIEW, link: false, filename: 'yield-to-call.csv',
        getRows: () => reportRows(lastReport),
    });
    mount.querySelectorAll('#ytc-price, #ytc-coupon, #ytc-mat, #ytc-call, #ytc-callprice, #ytc-freq').forEach(el => {
        el.addEventListener('input', debounce(compute, 250));
    });
    mount.querySelector('#ytc-run').addEventListener('click', compute);
    compute();
}

function reportRows(r) {
    if (!r || !r.valid) return [];
    return [
        ['metric', 'value'],
        ['ytm_pct', r.ytm_pct == null ? '' : r.ytm_pct],
        ['ytc_pct', r.ytc_pct == null ? '' : r.ytc_pct],
        ['ytw_pct', r.ytw_pct == null ? '' : r.ytw_pct],
        ['current_yield_pct', r.current_yield_pct],
        ['premium_to_par_pct', r.premium_to_par_pct],
        ['verdict', r.verdict],
    ];
}

function renderResult(mount, r, body) {
    const result = mount.querySelector('#ytc-result');
    if (!r.valid) {
        result.innerHTML = `<p class="muted">Enter a positive price.</p>`;
        return;
    }
    const verdictCls = r.verdict === 'YTC' ? 'neg' : 'pos';
    const isPremium = r.is_premium;
    const callRisk = isPremium
        ? '<strong>High call risk</strong> — bond trades above par. Issuer benefits by refinancing at lower rates and paying par. Plan on YTW.'
        : '<strong>Low call risk</strong> — bond trades at/below par. Issuer has no economic reason to call. YTM likely realized.';
    // The three yields side by side — the decision is the lowest (yield-to-worst).
    const chart = enh.svgBarChart([
        { label: 'YTM', value: r.ytm_pct || 0 },
        { label: 'YTC', value: r.ytc_pct || 0 },
        { label: 'YTW', value: r.ytw_pct || 0 },
    ]);
    // Cash flow trace (assumes called) — a schedule of the entered terms.
    const freq = body.coupons_per_year;
    const cpn = 1000 * (body.coupon_rate_pct / 100) / freq;
    const callPrice = 1000 * body.call_price_pct / 100;
    const periods = Math.ceil(body.years_to_call * freq);
    let rows = '';
    for (let i = 1; i <= periods; i++) {
        const yrs = i / freq;
        const principal = i === periods ? callPrice : 0;
        rows += `<tr><td>${i}</td><td class="muted">${fmt(yrs, 2)}</td><td>$${fmt(cpn, 2)}</td><td>${principal > 0 ? '<strong>$' + fmt(principal, 2) + '</strong>' : '—'}</td><td>$${fmt(cpn + principal, 2)}</td></tr>`;
    }
    result.innerHTML = `
        <div class="cards">
            <div class="card"><div class="label">Yield to Maturity</div><div class="value">${fmt(r.ytm_pct, 3)}%</div><div class="muted small">If held ${fmt(body.years_to_maturity, 1)} yr</div></div>
            <div class="card"><div class="label">Yield to Call</div><div class="value">${fmt(r.ytc_pct, 3)}%</div><div class="muted small">If called yr ${fmt(body.years_to_call, 1)}</div></div>
            <div class="card"><div class="label">Yield to Worst</div><div class="value ${verdictCls}"><strong>${fmt(r.ytw_pct, 3)}%</strong></div><div class="muted small">min(YTM, YTC) = ${esc(r.verdict)}</div></div>
            <div class="card"><div class="label">Current yield</div><div class="value">${fmt(r.current_yield_pct, 3)}%</div><div class="muted small">Coupon / price (ignores principal)</div></div>
            <div class="card"><div class="label">Premium / discount</div><div class="value ${isPremium ? 'neg' : 'pos'}">${isPremium ? '+' : ''}${fmt(r.premium_to_par_pct, 2)}</div><div class="muted small">vs par</div></div>
        </div>
        ${chart}
        <p class="muted small">${callRisk}</p>
        <h3 class="section-title">Cash flow trace (assumes called)</h3>
        <table class="trades" data-table-key="ytc-cf">
            <thead><tr><th>Period</th><th>Years</th><th>Coupon</th><th>Principal</th><th>Total</th></tr></thead>
            <tbody>${rows}</tbody>
        </table>
    `;
}

File: crates/traderview-core/src/yield_to_call.rs
//! Yield to maturity / call / worst for a callable bond. YTM is the IRR if held
//! to maturity (redeemed at par); YTC is the IRR if called at the call date and
//! call price; yield-to-worst is the lower of the two — what a prudent buyer of a
//! premium callable bond should underwrite to. Each yield is solved by bisection
//! on the price equation. Faithful port of the former client-side `solveYield`,
//! now Python-pinned and unit-tested. Pure compute, not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct YieldToCallInput {
    /// Clean price as a percent of par (106.5 = 106.5% of par).
    pub price_pct: f64,
    /// Annual coupon rate, percent.
    pub coupon_rate_pct: f64,
    pub years_to_maturity: f64,
    pub years_to_call: f64,
    /// Call price as a percent of par.
    pub call_price_pct: f64,
    #[serde(default = "default_freq")]
    pub coupons_per_year: u32,
}

fn default_freq() -> u32 {
    2
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct YieldToCallReport {
    pub ytm_pct: Option<f64>,
    pub ytc_pct: Option<f64>,
    pub ytw_pct: Option<f64>,
    pub current_yield_pct: f64,
    pub premium_to_par_pct: f64,
    pub is_premium: bool,
    /// "YTC", "YTM", or "TIE" — which yield is the worst (binds).
    pub verdict: String,
    pub valid: bool,
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

// Bisection on the annualized yield that prices the cash flows to `price`.
// Mirrors the original JS solver bracket [-0.20, 0.50] with hi-expansion.
fn solve_yield(price: f64, cpn: f64, redemption: f64, n: f64, freq: f64) -> Option<f64> {
    let f = |y: f64| -> f64 {
        let r = y / freq;
        if r.abs() < 1e-12 {
            return cpn * n + redemption - price;
        }
        let pw = (1.0 + r).powf(n);
        cpn * (1.0 - 1.0 / pw) / r + redemption / pw - price
    };
    let mut lo = -0.20;
    let mut hi = 0.50;
    let mut fl = f(lo);
    let mut fh = f(hi);
    if fl * fh > 0.0 {
        let mut i = 0;
        while i < 20 && fl * fh > 0.0 {
            hi *= 1.5;
            fh = f(hi);
            i += 1;
        }
        if fl * fh > 0.0 {
            return None;
        }
    }
    for _ in 0..200 {
        let mid = (lo + hi) / 2.0;
        let fm = f(mid);
        if fm.abs() < 1e-9 {
            return Some(mid);
        }
        if fl * fm < 0.0 {
            hi = mid;
            fh = fm;
        } else {
            lo = mid;
            fl = fm;
        }
    }
    let _ = fh;
    Some((lo + hi) / 2.0)
}

pub fn generate(i: &YieldToCallInput) -> YieldToCallReport {
    if i.price_pct <= 0.0 || i.coupons_per_year == 0 {
        return YieldToCallReport::default();
    }
    let par = 1000.0;
    let freq = i.coupons_per_year as f64;
    let price = par * i.price_pct / 100.0;
    let call_price = par * i.call_price_pct / 100.0;
    let cpn = par * (i.coupon_rate_pct / 100.0) / freq;

    let ytm = solve_yield(price, cpn, par, i.years_to_maturity * freq, freq);
    let ytc = solve_yield(price, cpn, call_price, i.years_to_call * freq, freq);
    let ytw = match (ytm, ytc) {
        (Some(m), Some(c)) => Some(m.min(c)),
        (Some(m), None) => Some(m),
        (None, Some(c)) => Some(c),
        (None, None) => None,
    };
    let verdict = match (ytm, ytc) {
        (Some(m), Some(c)) if c < m => "YTC",
        (Some(m), Some(c)) if c > m => "YTM",
        (Some(_), Some(_)) => "TIE",
        _ => "",
    };
    let current_yield = (cpn * freq) / price * 100.0;

    YieldToCallReport {
        ytm_pct: ytm.map(|y| round4(y * 100.0)),
        ytc_pct: ytc.map(|y| round4(y * 100.0)),
        ytw_pct: ytw.map(|y| round4(y * 100.0)),
        current_yield_pct: round4(current_yield),
        premium_to_par_pct: round4(i.price_pct - 100.0),
        is_premium: i.price_pct > 100.0,
        verdict: verdict.to_string(),
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> YieldToCallInput {
        YieldToCallInput {
            price_pct: 106.5,
            coupon_rate_pct: 5.5,
            years_to_maturity: 10.0,
            years_to_call: 3.0,
            call_price_pct: 100.0,
            coupons_per_year: 2,
        }
    }

    // Pins cross-checked against the original JS solveYield in Python.
    #[test]
    fn premium_callable_yields() {
        let d = generate(&base());
        assert!(close(d.ytm_pct.unwrap(), 4.6787));
        assert!(close(d.ytc_pct.unwrap(), 3.21));
        assert!(close(d.ytw_pct.unwrap(), 3.21));
        assert!(close(d.current_yield_pct, 5.1643));
        assert!(d.is_premium);
        assert!(close(d.premium_to_par_pct, 6.5));
        assert_eq!(d.verdict, "YTC");
    }

    #[test]
    fn at_par_ytm_equals_coupon() {
        // Priced at par with maturity = call → YTM ≈ coupon, no call advantage.
        let d = generate(&YieldToCallInput { price_pct: 100.0, years_to_call: 10.0, ..base() });
        assert!(close(d.ytm_pct.unwrap(), 5.5));
        assert!(!d.is_premium);
        assert_eq!(d.verdict, "TIE");
    }

    #[test]
    fn discount_bond_ytm_above_coupon() {
        let d = generate(&YieldToCallInput { price_pct: 92.0, ..base() });
        assert!(d.ytm_pct.unwrap() > 5.5);
        assert!(!d.is_premium);
        // A discount bond won't be called early, so YTM is the worst case.
        assert_eq!(d.verdict, "YTM");
    }

    #[test]
    fn zero_price_invalid() {
        let d = generate(&YieldToCallInput { price_pct: 0.0, ..base() });
        assert!(!d.valid);
    }
}
